### PR TITLE
Open acc

### PR DESCRIPTION
**Pull request by christophermaynard**
Finished GPU optimisation section. Needs merging onto master.

### CODE REVIEW

File: CompRep.tex [hunk 1]
```diff
@@ -1,6 +1,20 @@
 \documentclass[11pt]{article}
 \usepackage{graphicx}
 \usepackage{cite}
+\usepackage{xcolor}
+\usepackage{listings}
+
+\def\UrlBreaks{\do\/\do-}
+\lstset{language=[08]Fortran,
```
Comment (samcusworth) [inline CompRep.tex]: I think this line is causing latex compile issues on the linux desktop:
"
The following languages weren't loadable:
    ,fortran$08
"

File: PA.tex [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+\subsection{NVIDIA GPU}
+The LFRic-microbenchmark suite~\cite{lfric-microbenchmarks} allows
+optimisations for individual kernels to be explored without having to
+compile and run the whole code. This is particularly useful when
+targeting a programming model such as OpenACC for GPU architecture
+where the Portland group compiler cannot compile the Fortran 2003 used 
+in the main code.
+
+The local matrix assembly (LMA) matrix-vector kernel had been
+previously ported to run on GPUs by inserting Open ACC directives in
```
Comment (samcusworth) [inline PA.tex]: Here and throughout: "Open ACC" to "OpenACC" (no space)

File: PA.tex [hunk 3]
```diff
@@ -0,0 +1,173 @@
+\section{Optimisation for Processor Architectures
+\label{sec:pa}}
+
+\subsection{NVIDIA GPU}
+The LFRic-microbenchmark suite~\cite{lfric-microbenchmarks} allows
+optimisations for individual kernels to be explored without having to
+compile and run the whole code. This is particularly useful when
+targeting a programming model such as OpenACC for GPU architecture
```
Comment (samcusworth) [inline PA.tex]: "architecture" or "architectureS"?

File: PA.tex [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+Shown in listing~\ref{lst:LMA-orig} is a code fragment of the LMA
+kernel. It has an Open ACC directive \verb+!$acc loop vector+ before the
+outmost $k$ loop, so that each thread in a thread block will do different loop
+iterations. This code runs on a GPU, but not very efficiently. 
+
+The first optimisation was to fuse the two loops over \verb+ndf1+
+(lines 6-11 and 13-15) and replace the array \verb+lhs_e+ with a
+scalar. The scalar can stay resident in a register for each thread and
+avoids unnecessary memory accesses. This optimisation is call loop
```
Comment (samcusworth) [inline PA.tex]: "call" -> "called"

File: PA.tex [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+code is fairly well optimised.
+
+It should be noted that the problem size is too small, the data file
+supplied to NVIDIA had $864$ cells with $40$ layers. Whilst, a target
+local volume size of $12\times 12=144$ or $24\times 24=576$ is a
+reasonable size for a single core of a CPU, a similar size is too
+small to fully exploit the parallelism of GPU. In particular, to
+effectively exploit the SIMT parallelism with the vertical degrees of
+freedom, the vector size should be $128$ or more. $40$ levels is
+clearly to few. Worse, in this particular case the colouring data here
```
Comment (samcusworth) [inline PA.tex]: "to" -> "too"

File: PA.tex [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+The local matrix assembly (LMA) matrix-vector kernel had been
+previously ported to run on GPUs by inserting Open ACC directives in
+the PSy layer code to control the data transfer from host memory to
+device memory and to offload the kernel call to the GPU device. This
+Open ACC version was the starting point for Alan Gray from NVIDIA to
+to look at optimisations. Moreover, after discussions regarding 
+the PSyKAl API and the LFRic plans for how Open ACC will be
+implemented for a GPU version, NVIDIA believe that the plan and
+implementation will enable a succesful GPU implementation.
```
Comment (samcusworth) [inline PA.tex]: "succesful" -> "successful"

File: PA.tex [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+      lhs(map1(df)+k) = lhs(map1(df)+k) + lhs_e(df,k)
+    end do
+
+  end do 
+  !$
+\end{lstlisting}
+
+Shown in listing~\ref{lst:LMA-orig} is a code fragment of the LMA
+kernel. It has an Open ACC directive \verb+!$acc loop vector+ before the
+outmost $k$ loop, so that each thread in a thread block will do different loop
```
Comment (samcusworth) [inline PA.tex]: "outmost" -> "outermost"?

File: CompRep.tex (hunk 1 above)
Comment (samcusworth) [reply CompRep.tex]: Works if we get rid of "[08]"

File: PA.tex [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+          matrix(df,df2,ik)*x(map2(df2)+k)
+      end do
+    end do
+
+    do df = 1, ndf1
+      lhs(map1(df)+k) = lhs(map1(df)+k) + lhs_e(df,k)
+    end do
+
+  end do 
+  !$
```
Comment (samcusworth) [inline PA.tex]: As discussed, this sentinel isn't required in the final text

**Review by samcusworth — CHANGES_REQUESTED** [review]
Looks good. A couple of typos and a latex syntax thing for the linux desktop.

File: CompRep.tex (hunk 1 above)
Comment (christophermaynard) [reply CompRep.tex]: Yes, removing

File: PA.tex (hunk 2 above)
Comment (christophermaynard) [reply PA.tex]: Yes, done.

File: PA.tex (hunk 3 above)
Comment (christophermaynard) [reply PA.tex]: Let's go with plural.

File: PA.tex (hunk 4 above)
Comment (christophermaynard) [reply PA.tex]: Good spot.

File: PA.tex (hunk 5 above)
Comment (christophermaynard) [reply PA.tex]: Yes.

File: PA.tex (hunk 6 above)
Comment (christophermaynard) [reply PA.tex]: Yes.

File: PA.tex (hunk 7 above)
Comment (christophermaynard) [reply PA.tex]: Done.

File: PA.tex (hunk 8 above)
Comment (christophermaynard) [reply PA.tex]: Yup. Deleted.

**Review by samcusworth — APPROVED** [review]

**Pull request merged**